airbyte-integrations/connectors/source-fauna/source_fauna/spec.yaml

```diff
@@ -0,0 +1,210 @@
+documentationUrl: https://github.com/fauna/airbyte/blob/source-fauna/docs/integrations/sources/fauna.md
+connectionSpecification:
+  $schema: http://json-schema.org/draft-07/schema#
+  title: Fauna Spec
+  type: object
+  required:
+    - domain
+    - port
+    - scheme
+    - secret
+  additionalProperties: false
+  properties:
+    domain:
+      order: 0
+      type: string
+      title: Domain
+      description: >-
+        Domain of Fauna to query. Defaults db.fauna.com.
+        See <a href=https://docs.fauna.com/fauna/current/learn/understanding/region_groups#how-to-use-region-groups>the docs</a>.
+      default: "db.fauna.com"
+    port:
+      order: 1
+      type: integer
+      title: Port
+      description: Endpoint port.
+      default: 443
+    scheme:
+      order: 2
+      type: string
+      title: Scheme
+      description: URL scheme.
+      default: "https"
+    secret:
+      order: 3
+      type: string
+      title: Fauna Secret
+      description: Fauna secret, used when authenticating with the database.
+      airbyte_secret: true
+    collection:
```

Comment:
Every time a user must pass all collection they want to sync with every field and data type?

Reply:
This must be executed in Airbyte code, remove it from the user responsability.

Reply:
No, this is not how this works.

Fauna stores documents. By default all fields in the user's document will be exported.

However, if the user wishes to elevate certain _sub-fields_ of their document to top-level columns in the destination they can do so with these controls.

For collections, yes user's will have to setup a connector per collection. This is by design. As it enables users to specify different subfield controls as well as makes incremental sync a simple setup for the user (due to details of Fauna's internals).

Both of these choices were intentional and should give the user's a better experience.

Reply:
> However, if the user wishes to elevate certain sub-fields of their document to top-level columns in the destination they can do so with these controls.

For that the user can use dbt in destination to change data whenever they want. Following all other databases the user should input user/password/host/port and Airbyte connector will shows all table/collection/documents to user in the Connection Setup step. The user will have the option to select/unselect which one they don't want to sync.

I'll take a deeper look how incremental will work with Fauna. I'd just setup a test account 💯 

@misteryeo could you give your thoughts here?

Reply:
This does not preclude the usage of `dbt`. We thought it would be nice to include this for user's so they could set this once and forget it without having to build _any_ code at all.

Requiring them to use `dbt` forces them to maintain their own software.

With this we give user's a zero-touch option; which is what our customer's tell us they like about Fauna.

Reply:
I am going so second what @marcosmarxm said earlier. The need to specify collection details during connector setup makes this DB connector behavior different from all other DB connectors and also makes it harder to a user to sync multiple collections. Ideally, the connector will implement `discover` method that would return all collections and then user can chose which collections to sync.

On the second part of the comment about selecting certain fields and elevating certain sub-fields - this is a neat feature and we would eventually want to have it for all DB connectors. We plan to add the ability to select which fields to sync at the platform level (https://github.com/airbytehq/airbyte/issues/2227). And we also plan to make normalization more flexible. However, if you keep the current implementation in this connector, it will not be compatible with this feature in the future and will become hard to maintain. So, my suggestion is to not offer this feature at this point and to make the Fauna source connector behave more like other Database connectors. Specifically:

- let the user see all available collections and select the ones that the user wants to sync
- let the user see all available fields and add the code, which will read the field list from the platform as part of catalog. This way, when field filtering feature is implemented, this connector will get it for free

Reply:
We're experimenting with this internally.

One thing to note is that fauna has no schema, so the best we can do is provide a stream which just has a couple fields:
- `ref` The document's unique ID
- `ts` The timestamp of the most recent modification to the document
- `ttl` The time to live for the document
- `data` A JSON blob containing all other data for the document.

This format is difficult to work with in the destination, so it will require users to use DBT. If you'd like us to avoid the additionalColumns configuration, then this is the best we can do.

Additionally, this will not be fixed by #2227, as that only looks like that will allow field filtering. We don't have a list of fields at all, so this won't change things. Also, having the additionalColumns config will still be compatible with #2227, as you will still need that configuration in order for the source to know about the columns in the first place.

Reply:
How do other sources handle this? There are a number of schemaless databases, and the sources need to produce some type of schema for these databases.

Any recommendations on how we should produce a schema would be appreciated! Having a single `data` field would be clunky, and that seems like a last resort to me.

Reply:
Generally, in Airbyte's framework, sources do not care about destinations and vice versa. Right now, we do not have a good solution for normalizing schemaless data at destinations, especially if a destination is a SQL database, not a document store. There are two options here:

1. Read a sample of documents and report a set of fields in those documents as the schema. This is what MongoDB source does (see the code [here](https://github.com/airbytehq/airbyte/blob/f5bfe031eb04aa81614a34382e9c72a2899c840d/airbyte-db/db-lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java#L203))
2. Do what you are suggesting (return `ref` as a PK, `ts` as a cursor, `data` as a column that contains JSON representation of all fields)

First solution would make it easier for users who have similar documents. Second solution would be more flexible, but will require that users write additional normalization steps, which is not something we allow right now in the Cloud, but is on our roadmap. Do you know what is the more prevalent use pattern for Fauna customers?

Reply:
We've updated the connector to export four columns: data, ref, ts, and ttl. These will contain all the data for the Fauna document.

Sorry this took so long, we needed to have some internal discussion about how to proceed for this one.